[xcvrd] DPB support on platforms with CmisManagerTask enabled

ishidawataru opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->


#### Description

This commit fixes DPB support with CMIS transceivers.

CmisManagerTask's `port_dict` must be updated according to the port add/remove events.
This commit removes the `port_mapping` field from CmisManagerTask as
`port_mapping` was mostly used just for storing `asic_id` information
and that can be simply done by `port_dict` instead.

Added a helper method `get_asic_id()` method to CmisManagerTask for
getting `asic_id` from `logical_port`.

#### Motivation and Context

fixes https://github.com/sonic-net/sonic-buildimage/issues/18893

#### How Has This Been Tested?

I tested on the VS environment.

#### Additional Information (Optional)

mihirpat1 commented:
@ishidawataru Can you please fix the code coverage check failure?

ishidawataru commented:
@mihirpat1 I'm still communicating with the original reporter of this issue to check the fix actually works. After that I'll fix the code coverage failure.

ishidawataru commented:
@mihirpat1 Confirmed that the fix resolved the issue. Also, I fixed the coverage test failure. Please review.

prgeor reviewed (changes requested):
@ishidawataru There is not enough information for me to understand the problem. As long as CONFIG_DB port table has the right number of lanes and Speed specified, Xcvrd must be able to select the best matching application from the list of advertised application by the module. Do you see any issue in DPB port deletion path or DPB port creation path?

ishidawataru commented:
@prgeor This PR fixes https://github.com/sonic-net/sonic-buildimage/issues/18893. It is not about the application selection. The current `CmisManagerTask` implementation doesn't delete the port info from `port_dict` when ports are deleted due to DPB configuration, which causes a crash.

longhuan-cisco commented on sonic-xcvrd/xcvrd/xcvrd.py:
```diff
@@ -896,7 +897,13 @@ def on_port_update_event(self, port_change_event):
         else:
+            # PORT_DEL event for the same lport happens 3 times because
+            # we are subscribing to CONFIG_DB, STATE_DB|TRANSCEIVER_INFO, and STATE_DB|PORT_TABLE.
+            # We only handle the first one and ignore the rest.
```
Instead of introducing a new attribute `deleted_ports` and its corresponding update logic, you may be able to simply rely on port_change_event's existing `db_name`(/`table_name`), and only proceed for the case of db_name is CONFG_DB 
https://github.com/sonic-net/sonic-platform-daemons/blob/bf865c6b711833347d3c57e9d84cd366bcd1b776/sonic-xcvrd/xcvrd/xcvrd_utilities/port_event_helper.py#L30-L31

ishidawataru replied:
https://github.com/sonic-net/sonic-platform-daemons/pull/500/files#diff-fc70c43977a1ce57f3a6e9d7d09cb7ea5204f52b3adf3162936205e5680efb2fR899-R902

@longhuan-cisco  Thank you for the suggestion. I was able to fix the issue without introducing `deleted_ports`.

I initially added `deleted_ports` because the CmisManagerTask handles port events in two stages.

The first stage is done in `on_port_update_event` via   [`port_change_observer.handle_port_update_event()`](https://github.com/sonic-net/sonic-platform-daemons/pull/500/files#diff-fc70c43977a1ce57f3a6e9d7d09cb7ea5204f52b3adf3162936205e5680efb2fR1313) and the second stage is done after returning from `port_change_observer.handle_port_update_event()` by [`iterating over port_dict`](https://github.com/sonic-net/sonic-platform-daemons/pull/500/files#diff-fc70c43977a1ce57f3a6e9d7d09cb7ea5204f52b3adf3162936205e5680efb2fR1315).

If we delete the port info from `port_dict` in the first stage, the second stage cannot handle the port.
By introducing `deleted_ports`, the origin code was deleting the port info from `port_dict` after finishing the second stage.
However, it looks like currently, the second stage is not doing anything for the removed port. so deleting the port info in the first stage is safe.

Though this works, I think the current two-stage design is bug-prone, because we might want to add some handling(e.g. de-initialize) for the removed port in the second stage.
If you agree, I'd like to refactor the current design in a different PR.

longhuan-cisco commented on sonic-xcvrd/xcvrd/xcvrd.py:
```diff
@@ -1621,6 +1623,17 @@ def handle_cmis_state_machine(self, lport, info, is_fast_reboot):
+            self.handle_cmis_state_machine(lport, info, is_fast_reboot)
+
+        for event in self.deleted_ports.values():
+            self.port_dict.pop(event.port_name)
```
Would it make the flow/logic simpler by just directly deleting `self.port_dict[lport]` entry at the PORT_DEL event of CONFIG_DB in `on_port_update_event()`?

ishidawataru replied:
Done.

I explained the background of this change [here](https://github.com/sonic-net/sonic-platform-daemons/pull/500#discussion_r1762172401)

longhuan-cisco commented on sonic-xcvrd/xcvrd/xcvrd.py:
```diff
@@ -1300,354 +1302,360 @@ def wait_for_port_config_done(self, namespace):
-                        self.port_dict[lport]['appl'] = 0
-                        self.port_dict[lport]['host_lanes_mask'] = 0
-                    continue
+    def handle_cmis_state_machine(self, lport, info, is_fast_reboot):
```
The change here seems to be refactoring today's cmis_mgr's `task_worker` function by splitting it into sub functions. I'm assuming the flow/logic of cmis_mgr's task_worker doesn't get changed, it's mostly just splitting functions. But this makes the number of LOC (Line of change) very big. I feel It's a little bit difficult for other people to distinguish the actual change of DPB fix from the entire change set of this PR.

Thus, I feel it might be better to put this refactoring change set in a separate PR. It will also be less of a work for release mgr to revert in case of anything broken. Just my 2 cents, not something must-have.
Not sure how @mihirpat1 or Prince feel about this.

mihirpat1 replied:
@longhuan-cisco Agreed that splitting the current changeset into 2 separate PRs will be more helpful.

ishidawataru replied:
@longhuan-cisco @mihirpat1 I'm sorry for the late response. Thanks for the suggestion.

I'll move the refactoring commit to a separate PR and only include the DPB fix in this PR.

ishidawataru commented:
@longhuan-cisco @mihirpat1 I updated the PR based on @longhuan-cisco 's suggestion. Please review.

I used force-push and removed the original commits. I backed up the original branch [here](https://github.com/ishidawataru/sonic-platform-daemons/tree/fix-xcvrd-dpb-bak).

chiourung commented on sonic-xcvrd/xcvrd/xcvrd.py:
```diff
@@ -896,8 +895,13 @@ def on_port_update_event(self, port_change_event):
+        # To only handle the first one, check if the lport exists in the port_dict.
+        elif lport in self.port_dict:
             self.update_port_transceiver_status_table_sw_cmis_state(lport, CMIS_STATE_REMOVED)
+            self.port_dict.pop(lport)
```
If a transceiver is removed (not doing port breakout), the port would be removed from port_dict. When the transceiver is reattached to the port, there would be no port-related information in port_dict. I think the port should only be removed from port_dict after receiving a delete CONFIG_DB|PORT event and update cmis state after receiving a delete STATE_DB|TRANSCEIVER_INFO event.

ishidawataru replied:
Thank you for the comment. You're right. I'll update the code as you suggested.

ishidawataru replied:
@chiourung I fixed the issue with this commit. https://github.com/sonic-net/sonic-platform-daemons/pull/500/commits/954e9cb8f39554c2770c030c796c39feb667c194

Could you take a look?

ishidawataru commented:
@prgeor Hi, just a gentle ping to check if I need to do something to get this PR merged.

mssonicbld commented:
/azp run

ishidawataru commented:
@prgeor I fixed the conflicts. Please take a look.

prgeor commented:
@ishidawataru will you be able to test this on a real hardware with breakout enabled and do insertion and removal of the optics?

ishidawataru commented:
@prgeor Sorry, I don’t have access to real hardware to test this PR. I performed insertion and removal testing using the xcvr emulator in last October. I can redo those tests and provide the logs if needed.

longhuan-cisco commented on sonic-xcvrd/xcvrd/xcvrd.py:
```diff
@@ -681,8 +680,45 @@ def on_port_update_event(self, port_change_event):
+            if lport in self.port_dict:
+                self.update_port_transceiver_status_table_sw_cmis_state(lport, CMIS_STATE_REMOVED)
+
+            if port_change_event.db_name == 'CONFIG_DB' and port_change_event.table_name == 'PORT':
```
would it make the code more robust if we also put line 719-720 protected by line 717 `if lport in self.port_dict:` to avoid `self.port_dict.pop(lport)` throwing KeyError in the case of `lport` not present in port_dict? Though not sure if this rare case can actually happen.

ishidawataru replied:
Sure, I'll fix it as you mentioned.
I just thought if the case is not expected to happen, it's better to treat it as an error instead of silently ignoreing it.
By putting line 720-721 protected by line 717, we won't be able to see if this case is happening or not.

longhuan-cisco replied:
I see your concern, it makes sense to let the exception be thrown to not ignore error. 

The case might not always be an error case. One non-error case I can think of is: If xcvrd receives duplicate DEL events (though it's not very unlikely using today's [PortChangeObserver](https://github.com/sonic-net/sonic-platform-daemons/blob/ba47670352b1778ad3732d27c646e9a1f4b9c49b/sonic-xcvrd/xcvrd/xcvrd_utilities/port_event_helper.py#L155)), protecting it will help xcvrd survive.

It's your choice whether to protect it or leave it as is. I'm fine with either option.

prgeor commented on sonic-xcvrd/xcvrd/xcvrd.py:
```diff
@@ -2138,20 +2167,13 @@ def init(self):
-            # in this sense, we treat it as a part of new platform api.
-            # we have already moved sfputil to sonic_platform_base
-            # which is the root of new platform api.
-            platform_sfputil = sonic_platform_base.sonic_sfp.sfputilhelper.SfpUtilHelper()
```
@longhuan-cisco 
1. We want to get rid of _wrapper_xxx  as it was applicable when we needed to support platforms using API < v2.0. All platforms by now should be moving to Platform API v2.0.
2. sonic_platform_base.sonic_sfp is already deprecated.

longhuan-cisco replied:
got it, thanks for the info.

longhuan-cisco commented:
@prgeor 

I have done OIR tests with the latest diff. 

Tried below sequence on 400G AOC:
1. Move from non-breakout mode (1x400G) to 4x100G breakout mode via DPB CLI
2. Perform OIR
3. Move back to non-breakout mode (1x400G) via DPB CLI
4. Perform OIR

No xcvrd crash. link comes up fine with module ready state. `show interface` CLI shows correct interface entries.
